librz/core/cdebug.c

```diff
@@ -251,19 +238,71 @@ RZ_IPI bool rz_core_debug_reg_set(RzCore *core, const char *regname, ut64 val, c
 		rz_reg_set_value(core->dbg->reg, r, val);
 	}
 	rz_debug_reg_sync(core->dbg, RZ_REG_TYPE_ANY, true);
-	rz_core_debug_regs2flags(core, bits);
+	rz_core_debug_regs2flags(core);
 	return true;
 }
 
-RZ_IPI bool rz_core_debug_reg_list(RzCore *core, int type, int size, PJ *pj, int rad, const char *use_color) {
+static bool foreach_reg_cb(RzIntervalNode *node, void *user) {
+	RzRegItem *from_list = user;
+	RzRegItem *from_tree = node->data;
+	if (from_list == from_tree) {
+		return true;
+	}
+	// Check if from_list is covered entirely by from_tree, but is also smaller than it.
+	// We already know that
+	//   from_tree->offset <= from_list->offset < from_tree->offset + from_tree->size
+	if (from_list->offset + from_list->size > from_tree->offset + from_tree->size) {
+		// from_list expands beyond from_tree, so it's not covered
+		return true;
+	}
+	if (from_list->offset + from_list->size == from_tree->offset + from_tree->size) {
+		// they end at the same position, so it is covered entirely, but is it also smaller?
+		if (from_list->offset == from_tree->offset) {
+			// nope
+			return true;
+		}
+	}
+	// from_list ends before from_tree, so it is covered and smaller
+	return false;
+}
+
+/// Filter out all registers that are smaller than but covered entirely by some other register
+static RZ_OWN RzList *regs_filter_covered(RZ_BORROW const RzList /* <RzRegItem> */ *regs) {
```

Comment:
We should convert `regs` to vector in the future btw.

Reply:
Maybe. `RzList` isn't too bad here though since the number of regs should generally be well-controlled.